File: Contents/AutoLayout_Adaptivity/AutoLayout_Adaptivity/Examples/ViewControllerB.m
//
//  ViewControllerC.m
//  AutoLayout_Adaptivity
//
//  Created by Kwan Hyun Son on 2023/05/10.
//

#import "ViewControllerB.h"

@interface BackView : UIView
@end
@implementation BackView
+ (Class)layerClass {
    return [CAShapeLayer class];
}
- (void)awakeFromNib {
    [super awakeFromNib];
    self.backgroundColor = [UIColor clearColor];
}
- (void)layoutSubviews {
    [super layoutSubviews];
    [self updatePath];
}
- (void)updatePath {
    CAShapeLayer *shapeLayer = (CAShapeLayer *)self.layer;
    NSInteger length = self.layer.bounds.size.width;
    UIBezierPath *bezierPath = [UIBezierPath bezierPathWithRect:self.layer.bounds];
    [bezierPath moveToPoint:CGPointZero];
    [bezierPath addLineToPoint:CGPointMake(length, length)];
    [bezierPath moveToPoint:CGPointMake(0.0, length)];
    [bezierPath addLineToPoint:CGPointMake(length, 0.0)];
    [bezierPath closePath];
    
    shapeLayer.path = bezierPath.CGPath;
    shapeLayer.strokeColor = [UIColor blackColor].CGColor; // dashed line의 색깔.
    shapeLayer.fillColor  = [UIColor clearColor].CGColor;
    shapeLayer.lineCap = kCALineCapButt;
    shapeLayer.lineDashPattern = @[@(5), @(5)];
    shapeLayer.lineWidth = 2.0;  // dased line의 굵기
}
@end

@interface ViewControllerB ()
@property (weak, nonatomic) IBOutlet UIView *targetView;
@property (nonatomic) IBOutlet NSLayoutConstraint *centerXConstraint; // strong으로 잡아야 active 전환에서 문제를 방지한다.
@property (nonatomic) IBOutlet NSLayoutConstraint *centerYConstraint;
@end

@implementation ViewControllerB

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.targetView.translatesAutoresizingMaskIntoConstraints = NO;
    self.targetView.layer.cornerRadius = 25.0;
    UIBarButtonItem *rightBarButtonItem =
    [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemRefresh
                                                  target:self
                                                  action:@selector(runAction:)];
    rightBarButtonItem.tintColor = [UIColor blackColor];
    self.navigationItem.rightBarButtonItem = rightBarButtonItem;
}


#pragma mark - Actions
- (void)runAction:(UIBarButtonItem *)sender {
    sender.enabled = NO;
    [UIView animateKeyframesWithDuration:1.0
                                   delay:0.0
                                 options:UIViewKeyframeAnimationOptionAutoreverse
                              animations:^{
        [UIView addKeyframeWithRelativeStartTime:0.0
                                relativeDuration:0.25
                                      animations:^{
            self.centerXConstraint.active = NO;
            self.centerYConstraint.active = NO;
            self.centerXConstraint = [self.targetView.centerXAnchor constraintEqualToAnchor:self.targetView.superview.leadingAnchor];
            self.centerYConstraint = [self.targetView.centerYAnchor constraintEqualToAnchor:self.targetView.superview.topAnchor];
            self.centerXConstraint.active = YES;
            self.centerYConstraint.active = YES;
            [self.view layoutIfNeeded];
        }];
        [UIView addKeyframeWithRelativeStartTime:0.25
                                relativeDuration:0.25
                                      animations:^{
            self.centerXConstraint.active = NO;
            self.centerYConstraint.active = NO;
            self.centerXConstraint = [self.targetView.centerXAnchor constraintEqualToAnchor:self.targetView.superview.trailingAnchor];
            self.centerYConstraint = [self.targetView.centerYAnchor constraintEqualToAnchor:self.targetView.superview.topAnchor];
            self.centerXConstraint.active = YES;
            self.centerYConstraint.active = YES;
            [self.view layoutIfNeeded];
        }];
        [UIView addKeyframeWithRelativeStartTime:0.5
                                relativeDuration:0.5
                                      animations:^{
            self.centerXConstraint.active = NO;
            self.centerYConstraint.active = NO;
            self.centerXConstraint = [self.targetView.centerXAnchor constraintEqualToAnchor:self.targetView.superview.centerXAnchor];
            self.centerYConstraint = [self.targetView.centerYAnchor constraintEqualToAnchor:self.targetView.superview.centerYAnchor];
            self.centerXConstraint.active = YES;
            self.centerYConstraint.active = YES;
            [self.view layoutIfNeeded];
        }];
    }
                              completion:^(BOOL finished) {
        sender.enabled = YES;
    }];
}
@end

File: Contents/MGURegularPolygon/HexagonalWallpaper/Examples/ViewControllerB.swift
//
//  ViewControllerB.swift
//  HexagonalWallpaper
//
//  Created by Kwan Hyun Son on 2023/05/10.
//

import UIKit
import GraphicsKit
import IosKit

final class ViewControllerB: UIViewController {
    
    // MARK: - Property
    @IBOutlet private weak var centerView: UIView!  // 200.0 X 200.0
    @IBOutlet private weak var slider: UISlider!
    private var shapeLayer1 = CAShapeLayer()
    private var shapeLayer2 = CAShapeLayer()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        title = "동적으로 확인"
        centerView.backgroundColor = UIColor.white.withAlphaComponent(0.0)
        centerView.layer.cornerRadius = 100.0
        centerView.layer.borderColor = UIColor.black.cgColor
        centerView.layer.borderWidth = 0.5

        view.layer.addSublayer(shapeLayer1)
        shapeLayer1.frame = view.bounds
        shapeLayer1.lineWidth = 0.5
        shapeLayer1.strokeColor = UIColor.red.cgColor
        shapeLayer1.fillColor = UIColor.clear.cgColor
        
        view.layer.addSublayer(shapeLayer2)
        shapeLayer2.frame = view.bounds
        shapeLayer2.lineWidth = 0.5
        shapeLayer2.strokeColor = UIColor.red.cgColor
        shapeLayer2.fillColor = UIColor.clear.cgColor
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        shapeLayer1.frame = view.bounds
        shapeLayer2.frame = view.bounds
        let center = CGPoint(x: view.bounds.size.width / 2.0, y: view.bounds.size.height / 2.0)
        var path = UIBezierPath.init(regularPolygonAtCenter: center, radius: 100.0, numberOfSides: 3)
        shapeLayer1.path = path.cgPath
        path = UIBezierPath.init(bigRegularPolygonAtCenter: center, radius: 100.0, numberOfSides: 6, rotateRatio: 0.0)
        shapeLayer2.path = path.cgPath
    }
    
    
    // MARK: - Actions
    @IBAction func segmentedControlValueChanged(_ sender: UISegmentedControl) {
        let n = sender.selectedSegmentIndex + 3
        let center = CGPoint(x: view.bounds.size.width / 2.0, y: view.bounds.size.height / 2.0)
        var path = UIBezierPath.init(regularPolygonAtCenter: center, radius: 100.0, numberOfSides: UInt(n))
        shapeLayer1.path = path.cgPath
    }
    
    @IBAction func sliderValueChanged(_ sender: UISlider) {
        let center = CGPoint(x: view.bounds.size.width / 2.0, y: view.bounds.size.height / 2.0)
        var path = UIBezierPath.init(bigRegularPolygonAtCenter: center,
                                     radius: 100.0,
                                     numberOfSides: 6,
                                     rotateRatio: CGFloat(sender.value))
        shapeLayer2.path = path.cgPath
    }
}
